src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java

```diff
@@ -914,6 +942,34 @@ private void checkRelationMembersAndIndexRelation(final Relation relation,
         }
     }
 
+    private <Member extends AtlasEntity> Supplier<Iterable<Member>> getCachingSupplier(
+            final Iterable<Member> source, final ItemType type)
+    {
+        final List<Set<Long>> memberIdentifiersIntersecting = new ArrayList<>();
+        // A supplier that will effectively cache all the members intersecting that polygon. This is
+        // thread safe because the cache is internal to the supplier's scope.
+        @SuppressWarnings("unchecked")
+        final Supplier<Iterable<Member>> result = () ->
+        {
+            if (memberIdentifiersIntersecting.isEmpty())
+            {
+                final Set<Long> intersecting = new HashSet<>();
+                memberIdentifiersIntersecting.add(intersecting);
+                return Iterables.stream(source).map(member ->
+                {
+                    intersecting.add(member.getIdentifier());
+                    return member;
+                }).collect();
+            }
+            else
+            {
+                return (Iterable<Member>) Iterables.stream(memberIdentifiersIntersecting.get(0))
```

Comment:
I'm confused about the purpose of the List wrapper around the Set here. Since this pattern limits the List to fetching only one set, why not just use the Set directly?

Reply:
Because the object that goes from outside to inside the lambda has to be final. I cannot make the set final as I use the fact it might not be there as a cause for initializing it. So wrapping it in a final list is a trick that might be ugly but works :) I am open to any other more elegant suggestion!

Reply:
hmm... wouldn't it be possible to just use the `isEmpty()` check for initializing? a final `Set` can still be modified via `addAll()`, so you could do something like
```
final Set<Long> intersecting = new HashSet<>();
@SuppressWarnings("unchecked")
final Supplier<Iterable<Member>> result = () -> {
     if(intersecting.isEmpty()) {
           intersecting.addAll(Iterables.asSet(source));
     }
     return (Iterable<Member>) Iterables.stream(intersecting);
}
```
Not sure that code compiles or works, just spitballing here

Reply:
I am also curious to know if what @adahn6 suggests works or not.